### DISCUSSION TITLE
Discussions and possible FAQ

### DISCUSSION THREAD

**Discussion by hpsaturn** [jgauchia/IceNav-v3#107] [Ideas]
I notice that some issues right now, are more questions or doubts than issues, better we should use this section, Github Discussions, for these kind of topics. What do you think @jgauchia and @zhjygit ?

**Comment by jgauchia** [comment]
No problem , better

**Comment by MB-SMART** [comment]
Dear @jgauchia,
I am very interested in learning more about the Icenav project. Could you please let me know the best way to contact you for further discussion? Is your email available, or do you have a preferred method of communication?
Thank you very much.

**Reply by jgauchia** [reply]
Hi, thank you.
Because there are a contributor, maybe better here or in GitHub discussions

**Reply by MB-SMART** [reply]
How did you extract useful data from OpenStreetMap (OSM) for the Icenav project? Are there specific tools or libraries you recommend for this process?

What methods or algorithms did you use to determine the best path from the starting point to the destination?

**Reply by jgauchia** [reply]
There are 2 metods to obtain data, (explained in readme)

- render Maps (PNG tiles) extracted with maperitive soft. No OSM data

- vectorial Maps (draw with icenav) using OSM_extract project link in readme too. Here extract some OSM data like polygons, and something else. In future i'll extract streetnames, and other OSM data

Algorithm is still in process but tinyGPS++ library has a function that indicates distance from starting to endpoint.
Current icenav project can't determine (for now) the best path, this is, actually doesn't navigate like Google Maps, etc... Now i'm working in get distance for a endpoint and indications (left, right, etc).

Some day I'll make navigation style like Google Maps or similar , but I think that it's very difficult for me

**Comment by zhjygit** [comment]
so cool. contract me via ***@***.***



At 2024-07-29 14:53:46, "migga" ***@***.***> wrote:

Dear @jgauchia,
I am very interested in learning more about the Icenav project. Could you please let me know the best way to contact you for further discussion? Is your email available, or do you have a preferred method of communication?
Thank you very much.

**Reply by MB-SMART** [reply]
"so cool. contract me via ***@***.***"
Your replay is shown like this. the email address is shown as stars. It is not shown correctly. Please provide a way for contacting you.
I've sent you an email to the address shown in your github profile (jgauchia) with the name "Mahdi" and my email address starts with numbers.
Looking forward to your response.
Thank you.

**Reply by zhjygit** [reply]
djstmail@163.com

**Reply by MB-SMART** [reply]
Thank you for your prompt response. I have sent you an email for a more detailed discussion, but I also wanted to ask a couple of specific questions here:

How did you extract useful data from OpenStreetMap (OSM) for the Icenav project? Are there specific tools or libraries you recommend for this process?

What methods or algorithms did you use to determine the best path from the starting point to the destination?

Your guidance on these points would be incredibly helpful.

Thank you again for your time and assistance.

Best regards

**Reply by zhjygit** [reply]
1.at first, use the osm, there are some bugs that you can debug it.
and then i use the tool in https://www.centmap.cn/, that is very easy, it is GUI operation not as the OSM(python scripts).
2.There project is cool. 
based on it , like many issues i started, in fact ,i want to share gps location on the screens, that is to say, you can see many gps arrows on the screens. now, i have finished two devices sharing locations, sometime later, i will test 4 devices.

**Comment by brightproject** [comment]
Hello mr. @jgauchia
I have a microcontroller `esp32-s3-r16-n8` and a display `ILI9488` dimention `480x320`

![esp32_s3_n16_r8](https://github.com/user-attachments/assets/b0361a33-698c-452c-baa4-ae781caac5d2)

I connect the display via the interface `8-bit parallel(8080)`.
I can get the image via the library [lvgl](https://github.com/lvgl/lvgl)

![esp32-s3-n16-r8-8-bit-lvgl](https://github.com/user-attachments/assets/b8c31684-3963-49d1-b80f-d3cb5dde3080)

and

![esp32-s3-n16-r8-8-bit-etft-spi](https://github.com/user-attachments/assets/7f8f2fa4-731c-4382-aa1f-5b1f96e0bf65)

[TFT_eSPI](https://github.com/Bodmer/TFT_eSPI).

`User_Setup_Select` below

```
#define ESP32_PARALLEL
// #define TFT_PARALLEL_8_BIT

#define DISABLE_ALL_LIBRARY_WARNINGS

// #define ILI9341_DRIVER // DB8 = TFT_D0, DB9 = TFT_D1, DB10 =TFT_D2, DB11 =TFT_D3, DB12 =TFT_D4, DB13 = DB5, DB14 = TFT_D6, DB15 = TFT_D7, IM2 = 0, IM1 = 0, IM0 = 1
#define ILI9488_DRIVER // DB0 = TFT_D0, DB1 = TFT_D1, DB2 =TFT_D2, DB3 =TFT_D3, DB4 =TFT_D4, DB5 = DB5, DB6 = TFT_D6, DB7 = TFT_D7, IM2 = 0, IM1 = 1, IM0 = 1

#define TFT_CS 5
#define TFT_DC 7

#define TFT_RST -1// in scheme +3.3V

#define TFT_WR 4
#define TFT_RD 10

#define TFT_D0 14
#define TFT_D1 13
#define TFT_D4 16
#define TFT_D5 3
#define TFT_D6 9
#define TFT_D7 8
```

I usually compile my code through platformio with these settings:
```
[env:esp32-s3-devkitc1-n16r8]

platform = espressif32 @ 6.6.0

board = esp32-s3-devkitc1-n16r8
framework = arduino
monitor_speed = 115200
monitor_port = COM6
upload_speed = 921600
upload_port = COM22
monitor_filters = esp32_exception_decoder

lib_deps = bodmer/TFT_eSPI@^2.5.43
```
I try to compile the your code [IceNav-v3](https://github.com/jgauchia/IceNav-v3/).
Compile process below:

>  *  Executing task in folder IceNav-v3: C:\Users\test\.platformio\penv\Scripts\platformio.exe run --environment ESP32S3_N16R8 
> 
> Processing ESP32S3_N16R8 (board: adafruit_metro_esp32s3; platform: espressif32; framework: arduino)
> -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
> CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/adafruit_metro_esp32s3.html
> PLATFORM: Espressif 32 (6.8.1) > Adafruit Metro ESP32-S3
> HARDWARE: ESP32S3 240MHz, 320KB RAM, 16MB Flash
> DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-builtin, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
> PACKAGES:
>  - framework-arduinoespressif32 @ 3.20017.0 (2.0.17)
>  - tool-esptoolpy @ 1.40501.0 (4.5.1)
>  - toolchain-riscv32-esp @ 8.4.0+2021r2-patch5
>  - toolchain-xtensa-esp32s3 @ 8.4.0+2021r2-patch5
> LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
> LDF Modes: Finder ~ chain, Compatibility ~ soft
> Found 57 compatible libraries
> Scanning dependencies...
> Dependency Graph
> |-- TinyGPSPlus @ 1.1.0
> |-- Time @ 1.6.1
> |-- lvgl @ 9.1.0
> |-- LovyanGFX @ 1.1.16
> |-- Timezone @ 1.2.4
> |-- StreamUtils @ 1.9.0
> |-- EasyPreferences @ 0.1.2
> |-- battery
> |-- bme
> |-- compass
> |-- gps
> |-- lvgl
> |-- power
> |-- settings
> |-- storage
> |-- tasks
> |-- tft
> |-- FS @ 2.0.0
> |-- SD @ 2.0.0
> |-- SPI @ 2.0.0
> |-- SPIFFS @ 2.0.0
> |-- WiFi @ 2.0.0
> |-- Wire @ 2.0.0

Building in release mode
I get an error:

> Compiling .pio\build\ESP32S3_N16R8\lib2b8\lvgl\core\lv_obj_property.c.o
> In file included from lib/tasks/tasks.hpp:19,
>                  from src/main.cpp:42:
> lib/cli/cli.hpp:10:10: fatal error: ESP32WifiCLI.hpp: No such file or directory
> 
> ************************************************************************
> * Looking for ESP32WifiCLI.hpp dependency? Check our library registry!
> *
> * CLI  > platformio lib search "header:ESP32WifiCLI.hpp"
> * Web  > https://registry.platformio.org/search?q=header:ESP32WifiCLI.hpp
> *
> ************************************************************************
> 
>  #include <ESP32WifiCLI.hpp>
>           ^~~~~~~~~~~~~~~~~~
> compilation terminated.
> Archiving .pio\build\ESP32S3_N16R8\lib693\libTime.a
> *** [.pio\build\ESP32S3_N16R8\src\main.cpp.o] Error 1
> ============================================================================================================= [FAILED] Took 60.56 seconds =============================================================================================================
> Environment    Status    Duration
> -------------  --------  ------------
> ESP32S3_N16R8  FAILED    00:01:00.556

> ======================================================================================================== 1 failed, 0 succeeded in 00:01:00.556 ========================================================================================================

I am an experienced user of the library `TFT_eSPI`, a beginner user of `lvgl`, and I have been seeing your code for the second day.
Could you tell me what I am doing wrong?
P.S. 
You have an error in platformio.ini, this line is not commented out.

> https://github.com/jgauchia/IceNav-v3/blob/b4ad95594daecf7c4c792ed0dcb8c515bc4a2ad8/platformio.ini#L48
-------------------------------------------------------------------------------------------------
Oh, my mistake ... this is the right line and it loads the library for working with `Wi-Fi`.
But overall the code still doesn't build, because I want to make an output via `TFT_eSPI`, since it supports the `8-bit parallel` interface for my display.
In your code there is a library `LGFX_TFT_eSPI`, which only supports `ILI9488` via `SPI`.
In any case, I am interested in getting to know your code, I will figure it out.
I hope for a pleasant discussion🙂

**Reply by jgauchia** [reply]
Hi .

I'm current working to bring support for ILI9488 in 8 bit mode, but i don't have this TFT.

ILI9488 now only supported in SPI mode and 16bit Mode

You can see that there are a open issue for it.

If you can give me the gpio that you use i can modify the devel branch and add support for 8 bit mode with lovyangfx (is 100% compatible with tft_espi) and if you want you can test It.

Your Tft has touch? What type?

**Reply by jgauchia** [reply]
Sorry, I see the pinout in your comment. 

If you want tomorrow I'll Try to add your screen

**Reply by jgauchia** [reply]
I've crated a new branch https://github.com/jgauchia/IceNav-v3/tree/173-add-support-to-ILI9488-8bit , tomorrow I'll try to add 8bit mode and if you want, test it.

Thank you

**Reply by brightproject** [reply]
> I've crated a new branch https://github.com/jgauchia/IceNav-v3/tree/173-add-support-to-ILI9488-8bit , tomorrow I'll try to add 8bit mode and if you want, test it.
> 
> Thank you

This is great news, it would be great to add `8-bit` parallel to `lovyangfx` for `ILI9488` display.
My display does not have a `touch` screen, and I will try to add `TFT_eSPI` work myself.
I will be able to verify your code on my device once you update it.

**Reply by jgauchia** [reply]
Ok I'll Try to implement It , maybe I,'ve some extra work with lovyangfx and no touch panel, but not problem.
When I finish , I told you for testing

**Reply by jgauchia** [reply]
Please note that the project uses the touch to navigate through the screens.
For a quick test and since lovyangfx needs touch, I'll pretend that it has an i2c touch even if it's not used to see if it works while I look for information to initialize it without it.

[EDIT]
I've found how to initializate the TFT without touch, no problem with this...

**Reply by jgauchia** [reply]
Hi @brightproject 

Please try https://github.com/jgauchia/IceNav-v3/tree/173-add-support-to-ILI9488-8bit branch.

Ensure that in platformio.ini modify the following build flags under ESP32S3_N16R8 environment:

Change 
```-D ILI9488_FT5x06_SPI=1```
with
```-D ILI9488_NOTOUCH_8B=1```

Comment following build flag:
```-D SPI_SHARED=1```

Review **hal.hpp** and define your TFT pinout correctly for 8Bit mode, D8 to D15 you can put -1 if you want
https://github.com/jgauchia/IceNav-v3/blob/173-add-support-to-ILI9488-8bit/include/hal.hpp#L121

If it's correct and works, you can see the main screen (something like this) but you can't navigate to the other screens

![compass](https://github.com/user-attachments/assets/6328e944-f2ac-4072-80fe-e3fc162bb179)


Please if you can, give me some feedback.. 
Thank you

**Reply by brightproject** [reply]
> If it's correct and works, you can see the main screen (something like this) but you can't navigate to the other screens

Good evening @jgauchia 
I am very grateful to you for your participation and work with the code.
I started testing the new code on the `ESP32` and a `480x320` TFT display with a controller `ILI9488` without a touchscreen.
First, I edited the necessary pins in the `hal.hpp` file.
To avoid conflicts with my assigned pins on the printed circuit board, I specified all pins except those needed for the display as `-1`.

> 
> #ifndef HAL_HPP
> #define HAL_HPP
> 
> 
> /**
>  * @brief GPS pin definition
>  *
>  */
>   extern const uint8_t GPS_TX = -1;
>   extern const uint8_t GPS_RX = -1;
> 
> /**
>  * @brief I2C pin definition
>  *
>  */
>   #define I2C_SDA_PIN -1
>   #define I2C_SCL_PIN -1
> 
> /**
>  * @brief SD Card pin definition
>  *
>  */
> 
>   extern const uint8_t SD_CS = -1;
>   extern const uint8_t SD_MISO = -1;
>   extern const uint8_t SD_MOSI = -1;
>   extern const uint8_t SD_CLK = -1;
> 
> 
> 
> /**
>  * @brief TFT 8/16 Bits parallel pin definition
>  *
>  */
> extern const uint8_t TFT_CS  = 5;
> extern const uint8_t TFT_RS  = 7; // DC
> extern const uint8_t TFT_RST = -1;
> 
> extern const uint8_t TFT_WR  = 4;
> extern const uint8_t TFT_RD  = 10;
> 
> // 8 bit Mode
> extern const uint8_t TFT_D0  = 14;
> extern const uint8_t TFT_D1  = 13;
> extern const uint8_t TFT_D2  = 12;
> extern const uint8_t TFT_D3  = 15;
> extern const uint8_t TFT_D4  = 16;
> extern const uint8_t TFT_D5  = 3;
> extern const uint8_t TFT_D6  = 9;
> extern const uint8_t TFT_D7  = 8;
> 
> // Extra 16 bit Mode
> extern const uint8_t TFT_D8  = -1;
> extern const uint8_t TFT_D9  = -1;
> extern const uint8_t TFT_D10 = -1;
> extern const uint8_t TFT_D11 = -1;
> extern const uint8_t TFT_D12 = -1;
> extern const uint8_t TFT_D13 = -1;
> extern const uint8_t TFT_D14 = -1;
> extern const uint8_t TFT_D15 = -1;
> 
> /**
>  * @brief TOUCH I2C pin definition
>  *
>  */
> extern const uint8_t TCH_I2C_PORT = -1;
> extern const uint8_t TCH_I2C_SDA  = -1;
> extern const uint8_t TCH_I2C_SCL  = -1;
> extern const uint8_t TCH_I2C_INT  = -1;
> 
> #endif

I specified the flag in `platformio.ini`
```
build_flags =
${common.build_flags}
-D ILI9488_NOTOUCH_8B=1
```
And began to build the firmware.
The first build was about `1.5 minutes`.
After that, I loaded the firmware into the device.
When I turned on the power, the display turned `black`
![2](https://github.com/user-attachments/assets/45e0ae7a-5f60-4cc4-84fa-195cc85b55e6)
and then began to `flicker`
![1](https://github.com/user-attachments/assets/0c5a87ec-d075-41ae-85f3-6aaa4491ab5d)
sometimes showing stripes and colored pixels.


https://github.com/user-attachments/assets/b38b8e14-5e4d-4384-82c0-d0d302134ab8


But the serial port console received information with errors

> rst:0x1 (POWERON),boot:0xa (SPI_FAST_FLASH_BOOT)
> SPIWP:0xee
> mode:DIO, clock div:1
> load:0x3fce3808,len:0x4bc
> load:0x403c9700,len:0xbd8
> load:0x403cc700,len:0x2a0c
> entry 0x403c98d0
> [   317][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   323][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   329][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   335][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   341][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   347][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   353][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   359][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   365][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   371][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   377][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   384][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   390][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   396][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   402][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   408][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   414][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   420][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   426][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   432][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   438][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   444][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   450][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   456][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   462][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   468][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   474][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   480][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   487][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   493][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   499][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   505][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   511][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   517][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   523][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   529][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [   535][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
> E (235) gpio: gpio_set_level(227): GPIO output gpio_num error
> [   542][E][esp32-hal-spi.c:178] spiAttachSCK(): HSPI Does not have default pins on ESP32S3!
> [   555][E][esp32-hal-spi.c:215] spiAttachMISO(): HSPI Does not have default pins on ESP32S3!
> [   563][E][esp32-hal-spi.c:254] spiAttachMOSI(): HSPI Does not have default pins on ESP32S3!
> [   572][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
> E (271) gpio: gpio_set_level(227): GPIO output gpio_num error
> E (272) gpio: gpio_set_level(227): GPIO output gpio_num error
> E (273) gpio: gpio_set_level(227): GPIO output gpio_num error
> E (776) gpio: gpio_set_level(227): GPIO output gpio_num error
> [  1083][E][sd_diskio.cpp:806] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
> E (791) gpio: gpio_set_level(227): GPIO output gpio_num error
> [  1633][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
> E (1332) gpio: gpio_set_level(227): GPIO output gpio_num error
> [  1638][E][storage.cpp:38] initSD(): SD Card Mount Failed
> [  2201][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
> E (1901) gpio: gpio_set_level(227): GPIO output gpio_num error
> [  2833][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
> E (2533) gpio: gpio_set_level(227): GPIO output gpio_num error
> [  3464][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
> E (3164) gpio: gpio_set_level(227): GPIO output gpio_num error
> 
> [ 15405][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [ 15412][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [ 15418][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [ 15424][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> [ 15430][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND
> 
> SSID is empty, please set a valid SSID into quotes
> 
> [ 15447][E][Preferences.cpp:50] begin(): nvs_open failed: NOT_FOUND

and in the end, the console loaded `IceNav :$`

![load_cli](https://github.com/user-attachments/assets/5b138c45-fe21-42f0-8e73-a3a29c44484e)

Please tell me, do you use `lvgl` to display graphics on your displays?
It is based on `tfte_SPI`, so why use hal to specify pins?
For me, this add-on or abstraction in the form of hal complicates working with the code.
After all, lvl is already a kind of add-on over `tft_eSPI`?
I began to figure out why my display flickers and there is no image.
In the file
`LGFX_TFT_eSPI.hpp`
all attributes are left for `9341`, is this a mistake?
```
using namespace lgfx::textdatum;
using namespace lgfx::attribute;
using namespace lgfx::ili9341_colors;
using namespace lgfx::tft_command;
```
I want to compare the file

> .pio\libdeps\ESP32S3_N16R8\LovyanGFX\src\lgfx\v1\panel\Panel_ILI948x.hpp

with the ILI9488 initialization file in the library

> .pio\libdeps\ESP32S3_N16R8\TFT_eSPI\TFT_Drivers\ILI9488_Defines.h

and

> .pio\libdeps\ESP32S3_N16R8\TFT_eSPI\TFT_Drivers\ILI9488_Init.h

It seemed to me that the registers in this file do not match `ILI9488`.

```
const uint8_t* getInitCommands(uint8_t listno) const override
    {
      static constexpr uint8_t list0[] =
      {
          CMD_PWCTR1,  2, 0x17,  // VRH1
                          0x15,  // VRH2
          CMD_PWCTR2,  1, 0x41,  // VGH, VGL
          CMD_VMCTR1,  3, 0x00,  // nVM
                          0x12,  // VCM_REG
                          0x80,  // VCM_REG_EN
          CMD_FRMCTR1, 1, 0xA0,  // Frame rate = 60Hz
          CMD_INVCTR,  1, 0x02,  // Display Inversion Control = 2dot
          CMD_DFUNCTR, 3, 0x02,  // Nomal scan
                          0x22,  // 5 frames
                          0x3B,
          CMD_ETMOD,   1, 0xC6,  // 
          CMD_ADJCTL3, 4, 0xA9,  // Adjust Control 3 
                          0x51,
                          0x2C,
                          0x82,

          CMD_GMCTRP1,15, 0x0F, 0x1F, 0x1C, 0x0C, 0x0F, 0x08, 0x48, 0x98,
                          0x37, 0x0A, 0x13, 0x04, 0x11, 0x0D, 0x00,

          CMD_GMCTRN1,15, 0x0F, 0x32, 0x2E, 0x0B, 0x0D, 0x05, 0x47, 0x75,
                          0x37, 0x06, 0x10, 0x03, 0x24, 0x20, 0x00,

          CMD_SLPOUT , 0+CMD_INIT_DELAY, 120,    // Exit sleep mode
          CMD_IDMOFF , 0,
          CMD_DISPON , 0+CMD_INIT_DELAY, 100,
          0xFF,0xFF, // end
      };
      switch (listno)
      {
      case 0: return list0;
      default: return nullptr;
      }
    }
  };
```
Anyway, the work has started and it's great that you are helping to understand the code.
Thanks🙂

**Reply by jgauchia** [reply]
Hi, 
I'm not using tft_espi, lovyangfx works equal, has other funcions like draw PNG, JPEG, etc native, and is a little optimized for esp32.

Yes , I'm using LVGL for all graphics except map. 

I'm using hal in order to configure different panels, pinouts, etc . 

I've succesfully initializate ILI9488 panel in SPI mode and 16bit Mode, using lovyangfx, this indicate that initialization code is correct.

Do you have any datasheet or info of your board?

**Reply by brightproject** [reply]
> Do you have any datasheet or info of your board?

The printed circuit board is homemade, it has an `esp32s3n16r8`, a connector for the fpc connector and an encoder.
I can run other examples without problems.

https://github.com/user-attachments/assets/8f2b2cd4-6b66-4f17-a550-5f71026d879c


I think I need to check the registers and compare the parameters with the display initialization in the `tfte_SPI`  library.

**Reply by jgauchia** [reply]
Ok, tomorrow I'll check again the code. I told you something when I'll finish.

**Reply by jgauchia** [reply]
Also, you can try to test with lower speed

Change **cfg.freq_write = 20000000;** in ILI9488_NOTOUCH_8B.hpp to lower speed like 10MHz 

There is an example for use lovyangfx alone 
https://github.com/lovyan03/LovyanGFX/issues/258

Maybe you can test It without LVGL or my code ,because it's possible that other components has conflicts.

**Reply by brightproject** [reply]
> Also, you can try to test with lower speed

Thanks, mr @jgauchia 🙂
I managed to run the code and the image appeared on the display.
I left the connection to the pins unchanged, the speed `cfg.freq_write = 20000000;` does not affect the work.
I had to comment out some tasks from the main.cpp:
```
powerOn();
// loadPreferences();
// initSD();
initSPIFFS();
initTFT();
// initGPS();
initLVGL();
```
First, the information about the microcontroller was loaded in yellow font

![6](https://github.com/user-attachments/assets/6f0ac3a7-287e-4307-8913-9cb9ab022629)

and then the compass screen, but it seemed like some elements were missing

![5](https://github.com/user-attachments/assets/c912330a-637e-410f-9142-2eb94e9d76b0)

I figured it out a little and loaded the pictures
`pio run --target uploadfs`

Still, the register values ​​are slightly different from `Tfte_spi`, but when replacing the register

> https://github.com/Bodmer/TFT_eSPI/blob/5793878d24161c1ed23ccb136f8564f332506d53/TFT_Drivers/ILI9488_Init.h#L76

```
writecommand(0xB6); // Display Function Control
writedata(0x02);
writedata(0x02);
writedata(0x3B);
```

In `LovyanGFX`

> https://github.com/lovyan03/LovyanGFX/blob/50a1800eb771ff8c028334d055ad37e1f12ec863/src/lgfx/v1/panel/Panel_ILI948x.hpp#L211
```
  CMD_DFUNCTR, 3, 0x02,  // Nomal scan
                  0x22,  // 5 frames
                  0x3B,
```

I got a mirrored image and returned everything as it was.

I have an encoder on my board, I'll try to use it to navigate the menu.
For some reason code
`initLvglTask();`
puts the esp32 into `panic mode`.

The loading is going well, I'll continue to figure it out.

![7](https://github.com/user-attachments/assets/5ccd3f63-ca9e-4ab1-9c47-a0e619cb76e9)

![8](https://github.com/user-attachments/assets/033abc5c-9a31-4a75-a7ad-d8808625a20f)

How do you write in the console, in the port monitor, through vskod and platformio I can’t type anything in the console?

![load_cli_2](https://github.com/user-attachments/assets/83884a4e-bb79-43ff-80b6-a9c6734cec91)

**Reply by jgauchia** [reply]
Please try to add this at platformio.ini if you use platformio console under board definition

```monitor_echo = yes```
```monitor_filters = send_on_enter```

Finally I've removed LVGL Task in devel branch , if you want I'll merge devel with this branch

**Comment by brightproject** [comment]
Your project is also interesting for another reason - you have mastered the `platformio` and its settings and configurations quite well. For example, how did you implement such information output?

![status](https://github.com/user-attachments/assets/4cf168c8-4a51-438b-a3b2-0b3c65c1271d)

**Reply by jgauchia** [reply]
This output? I think that it's platformio standard output

**Reply by brightproject** [reply]
> This output? I think that it's platformio standard output

No, this is some specific conclusion, I have a different conclusion.
![status_another](https://github.com/user-attachments/assets/41f2a39a-4516-4b0c-9d9f-b346447dbc83)
It is formed by some specific settings in the `platformio.ini` file.
Please show your file `ESP32S3_N16R8.json`

**Reply by jgauchia** [reply]
I don't have this file, only the json of the board definition.
This output was made when I made pio run in all the environments.

[esp32-s3-devkitc-1-n16r8v.json](https://github.com/user-attachments/files/16574097/esp32-s3-devkitc-1-n16r8v.json)

**Reply by brightproject** [reply]
> I don't have this file, only the json of the board definition. This output was made when I made pio run in all the environments.
> 
> [esp32-s3-devkitc-1-n16r8v.json](https://github.com/user-attachments/files/16574097/esp32-s3-devkitc-1-n16r8v.json)

I specify the name
`board = esp32-s3-devkitc-1-n16r8v`
as the name of the `JSON` file, but your names of the environment variables `ESP32_N16R4` are somehow not similar to the names of the `JSON` files.
Where does the mapping between the alias `ESP32_N16R4` and the `JSON` file name occur?
That's why I'm interested in your setup🙂

**Reply by jgauchia** [reply]
Environment is how you can define different combinations of board or settings, configs, etc. 

I use these names in order to describe the board, but it's possible any name

Later you specify the environment in pio run -e environment

In my project there are 3 board

- ESP32
- MAKERFABS ESP32S3
- ESP32S3

With the environment you can choose the board and compile for it

**Reply by jgauchia** [reply]
> > This output? I think that it's platformio standard output
> 
> No, this is some specific conclusion, I have a different conclusion.
> ![status_another](https://github.com/user-attachments/assets/41f2a39a-4516-4b0c-9d9f-b346447dbc83)
> It is formed by some specific settings in the `platformio.ini` file.
> Please show your file `ESP32S3_N16R8.json`
> 

Maybe plataformio does that if you have more than one environment

**Reply by brightproject** [reply]
> Maybe plataformio does that if you have more than one environment

Exactly right 🙂

```
[platformio]
description = BNO055
default_envs = genericSTM32F411CE

[env:genericSTM32F401CE]
platform = ststm32
board = blackpill_f401ce

[env:genericSTM32F411CE]
platform = ststm32
platform_packages = platformio/framework-arduinoststm32@^4.20801.240802

board = blackpill_f411ce
framework = arduino
```

![status_one_more_enviroment_board](https://github.com/user-attachments/assets/6584a8c9-7cd6-4ebd-b41e-4dc0b9246ce0)
Thx! @jgauchia

**Comment by Drworkstation** [comment]
Hello, my challenge in understanding this project is how to link the extracted GPS position to the downloaded maps. In other words, if the GPS shows position X, how do we know which part of the maps stored on the memory card to display?

**Reply by jgauchia** [reply]
Hi!
With extracted GPS position there are a litte function that convert the coordinates to the correspondent map folder and get full path of the PNG file .
In render maps there are a function that do it

``` getMapTile(double lon, double lat, uint8_t zoomLevel, int16_t offsetX, int16_t offsetY) ```

First I convert latitude and longitude to X and Y values that it's corresponds to TileX and TileY folders and disired zoom of openstreetmap tile format

```
uint32_t lon2tilex(double f_lon, uint8_t zoom)
uint32_t lat2tiley(double f_lat, uint8_t zoom)
```

All are mathematical functions you can see it at file **renderMaps.cpp**

Also you can see this link https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames where everything is explained

**Reply by Drworkstation** [reply]
Thanks🙏Another question: Could you please let me know when the latest version of the PCB will be available on GitHub? Also, how are the physical buttons driven when using a non-touch display?

**Reply by jgauchia** [reply]
Hi, 
I'm currently working in a prototype PCB, It will be available soon, when I do tests and some other improvements
(Remember, this is a hobby project, and I have to balance family, work, etc... hahaha)

At the moment IceNav doesn't use buttons, everything is touch-based.

Actually, like everything, we are used to using touch inputs like mobile phones, etc... I opted for this option, although the first versions, a long time ago, did have physical buttons.

It is true that there are already several people who want to use physical inputs such as buttons or encoders.
I don't rule out doing a workaround to create a derivative that allows this.
At the moment I don't have it planned in the development roadmap but I don't rule out doing it in parallel to the project.

Please, feel free to create an Issue for it if you want. (indicating what type of input you would like to have, buttons, encoders, etc.)

(please , sorry for my bad English :( )

**Reply by Drworkstation** [reply]
Thank you for your prompt response. I understand your busy schedule and am eagerly waiting for the release of the next version of the PCB👍. If I can be of any help, I would be happy if you let me know. 🙏

**Reply by jgauchia** [reply]
No problem.
Any help, input or whatever is always welcome (keep in mind that I have some knowledge limitations hehehe)

**Reply by Drworkstation** [reply]
Me too(very limited😅)

**Reply by zhjygit** [reply]
physical buttons,that is cool.
Operation of touch lcd is not good in some situations,like zoom in/out, physical buttons sometimes are more efficent.
Waiting for your updates.

**Reply by jgauchia** [reply]
Yes, and not... 
I'll take note implement hadware buttons in future version , for now, doesn't tell when... First I need to fix some bugs in hardware and implement new features at software
